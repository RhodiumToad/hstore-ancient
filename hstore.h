/*
 * $PostgreSQL: pgsql/contrib/hstore/hstore.h,v 1.8 2009/06/11 14:48:51 momjian Exp $
 */
#ifndef __HSTORE_H__
#define __HSTORE_H__

#include "fmgr.h"
#include "utils/array.h"

/* there is one of these for each key _and_ value */
/* the value points to the _end_ so that we can get the length
 * by subtraction from the previous entry.
 */

typedef struct
{
	uint32  entry;
} HEntry;

#define HENTRY_ISFIRST 0x80000000
#define HENTRY_ISNULL  0x40000000
#define HENTRY_POSMASK 0x3FFFFFFF

/* determined by the size of "endpos", though this is a bit academic
 * since currently varlenas (and hence both the input and the whole hstore)
 * have the same limit
 */
#define HSTORE_MAX_KEY_LEN 0x3FFFFFFF
#define HSTORE_MAX_VALUE_LEN 0x3FFFFFFF

typedef struct
{
	int32		vl_len_;		/* varlena header (do not touch directly!) */
	int4		size;
    /* array of HEntry follows */
} HStore;

#define HSHRDSIZE	(sizeof(HStore))
#define CALCDATASIZE(x, lenstr) ( (x) * 2 * sizeof(HEntry) + HSHRDSIZE + (lenstr) )

/* note multiple evaluations of x */
#define ARRPTR(x)		( (HEntry*) ( (HStore*)(x) + 1 ) )
#define STRPTR(x)		( (char*)(ARRPTR(x) + ((HStore*)x)->size * 2) )

/* note multiple/non evaluations */
#define HSE_ISFIRST(he_) (((he_).entry & HENTRY_ISFIRST) != 0)
#define HSE_ISNULL(he_) (((he_).entry & HENTRY_ISNULL) != 0)
#define HSE_ENDPOS(he_) ((he_).entry & HENTRY_POSMASK)
#define HSE_OFF(he_) (HSE_ISFIRST(he_) ? 0 : HSE_ENDPOS((&(he_))[-1]))
#define HSE_LEN(he_) (HSE_ISFIRST(he_)	\
					  ? HSE_ENDPOS(he_) \
					  : HSE_ENDPOS(he_) - HSE_ENDPOS((&(he_))[-1]))

#define HS_KEY(arr_,str_,i_) ((str_) + HSE_OFF((arr_)[2*(i_)]))
#define HS_VAL(arr_,str_,i_) ((str_) + HSE_OFF((arr_)[2*(i_)+1]))
#define HS_KEYLEN(arr_,i_) (HSE_LEN((arr_)[2*(i_)]))
#define HS_VALLEN(arr_,i_) (HSE_LEN((arr_)[2*(i_)+1]))
#define HS_VALISNULL(arr_,i_) (HSE_ISNULL((arr_)[2*(i_)+1]))

/* currently, these following macros are the _only_ places that rely
 * on internal knowledge of HEntry. Everything else should be using
 * the above macros.
 */

/* copy one key/value pair (which must be contiguous starting at
 * sptr_) into an under-construction hstore; dent_ is an HEntry*,
 * dbuf_ is the destination's string buffer, dptr_ is the current
 * position in the destination. lots of modification and multiple
 * evaluation here.
 */
#define HS_COPYITEM(dent_,dbuf_,dptr_,sptr_,klen_,vlen_,vnull_)			\
    do {																\
		memcpy((dptr_), (sptr_), (klen_)+(vlen_));						\
		(dptr_) += (klen_)+(vlen_);										\
		(dent_)++->entry = ((dptr_) - (dbuf_) - (vlen_)) & HENTRY_POSMASK; \
		(dent_)++->entry = ((((dptr_) - (dbuf_)) & HENTRY_POSMASK)		\
							 | ((vnull_) ? HENTRY_ISNULL : 0));			\
	} while(0)

/* add one key/item pair, from a Pairs structure, into an
 * under-construction hstore
 */
#define HS_ADDITEM(dent_,dbuf_,dptr_,pair_)								\
	do {																\
		memcpy((dptr_), (pair_).key, (pair_).keylen);					\
		(dptr_) += (pair_).keylen;										\
		(dent_)++->entry = ((dptr_) - (dbuf_)) & HENTRY_POSMASK;		\
		if ((pair_).isnull)												\
			(dent_)++->entry = ((((dptr_) - (dbuf_)) & HENTRY_POSMASK)	\
								 | HENTRY_ISNULL);						\
		else															\
		{																\
			memcpy((dptr_), (pair_).val, (pair_).vallen);				\
			(dptr_) += (pair_).vallen;									\
			(dent_)++->entry = ((dptr_) - (dbuf_)) & HENTRY_POSMASK;	\
		}																\
	} while (0)

/* finalize a newly-constructed hstore */
#define HS_FINALIZE(hsp_,count_,buf_,ptr_)							\
	do {															\
		int	buflen = (ptr_) - (buf_);								\
		if ((count_))												\
			ARRPTR(hsp_)[0].entry |= HENTRY_ISFIRST;				\
		if ((count_) != (hsp_)->size)								\
		{															\
			(hsp_)->size = (count_);								\
			memmove(STRPTR(hsp_), (buf_), buflen);					\
		}															\
		SET_VARSIZE((hsp_), CALCDATASIZE((count_), buflen));		\
	} while (0)

/* ensure the varlena size of an existing hstore is correct */
#define HS_FIXSIZE(hsp_,count_)											\
	do {																\
		int bl = (count_) ? HSE_ENDPOS(ARRPTR(hsp_)[2*(count_)-1]) : 0;	\
		SET_VARSIZE((hsp_), CALCDATASIZE((count_),bl));					\
	} while (0)

#define PG_GETARG_HS(x) ((HStore*)PG_DETOAST_DATUM(PG_GETARG_DATUM(x)))

typedef struct
{
	char	   *key;
	char	   *val;
	size_t		keylen;
	size_t		vallen;
	bool		isnull;
	bool		needfree;
} Pairs;

int			hstoreUniquePairs(Pairs * a, int4 l, int4 *buflen);
HStore *    hstorePairs(Pairs *pairs, int4 pcount, int4 buflen);

size_t		hstoreCheckKeyLen(size_t len);
size_t		hstoreCheckValLen(size_t len);

int         hstoreFindKey(HStore * hs, int *lowbound, char *key, int keylen);
Pairs *     hstoreArrayToPairs(ArrayType *a, int* npairs);

#define HStoreContainsStrategyNumber	7
#define HStoreExistsStrategyNumber		9
#define HStoreExistsAnyStrategyNumber	10
#define HStoreExistsAllStrategyNumber	11

/* defining HSTORE_POLLUTE_NAMESPACE=0 will prevent this; for now, we default
 * to on for the benefit of people restoring old dumps
 */
#ifndef HSTORE_POLLUTE_NAMESPACE
#define HSTORE_POLLUTE_NAMESPACE 1
#endif

#if HSTORE_POLLUTE_NAMESPACE
#define HSTORE_POLLUTE(newname_,oldname_) \
	PG_FUNCTION_INFO_V1(oldname_);		  \
	Datum oldname_(PG_FUNCTION_ARGS);	  \
	Datum newname_(PG_FUNCTION_ARGS);	  \
	Datum oldname_(PG_FUNCTION_ARGS) { return newname_(fcinfo); }
#else
#define HSTORE_POLLUTE(newname_,oldname_)
#endif

#endif   /* __HSTORE_H__ */
