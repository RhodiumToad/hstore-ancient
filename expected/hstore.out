--
-- first, define the datatype.  Turn off echoing so that expected file
-- does not depend on contents of hstore.sql.
--
SET client_min_messages = warning;
\set ECHO none
RESET client_min_messages;
set escape_string_warning=off;
--hstore;
select ''::hstore;
 hstore 
--------
 
(1 row)

select 'a=>b'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select ' a=>b'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select 'a =>b'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select 'a=>b '::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select 'a=> b'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select '"a"=>"b"'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select ' "a"=>"b"'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select '"a" =>"b"'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select '"a"=>"b" '::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select '"a"=> "b"'::hstore;
  hstore  
----------
 "a"=>"b"
(1 row)

select 'aa=>bb'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select ' aa=>bb'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select 'aa =>bb'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select 'aa=>bb '::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select 'aa=> bb'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select '"aa"=>"bb"'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select ' "aa"=>"bb"'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select '"aa" =>"bb"'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select '"aa"=>"bb" '::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select '"aa"=> "bb"'::hstore;
   hstore   
------------
 "aa"=>"bb"
(1 row)

select 'aa=>bb, cc=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>bb , cc=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>bb ,cc=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>bb, "cc"=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>bb , "cc"=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>bb ,"cc"=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>"bb", cc=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>"bb" , cc=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>"bb" ,cc=>dd'::hstore;
         hstore         
------------------------
 "aa"=>"bb", "cc"=>"dd"
(1 row)

select 'aa=>null'::hstore;
   hstore   
------------
 "aa"=>NULL
(1 row)

select 'aa=>NuLl'::hstore;
   hstore   
------------
 "aa"=>NULL
(1 row)

select 'aa=>"NuLl"'::hstore;
    hstore    
--------------
 "aa"=>"NuLl"
(1 row)

select '\\=a=>q=w'::hstore;
   hstore    
-------------
 "=a"=>"q=w"
(1 row)

select '"=a"=>q\\=w'::hstore;
   hstore    
-------------
 "=a"=>"q=w"
(1 row)

select '"\\"a"=>q>w'::hstore;
    hstore    
--------------
 "\"a"=>"q>w"
(1 row)

select '\\"a=>q"w'::hstore;
    hstore     
---------------
 "\"a"=>"q\"w"
(1 row)

select ''::hstore;
 hstore 
--------
 
(1 row)

select '	'::hstore;
 hstore 
--------
 
(1 row)

-- -> operator
select 'aa=>b, c=>d , b=>16'::hstore->'c';
 ?column? 
----------
 d
(1 row)

select 'aa=>b, c=>d , b=>16'::hstore->'b';
 ?column? 
----------
 16
(1 row)

select 'aa=>b, c=>d , b=>16'::hstore->'aa';
 ?column? 
----------
 b
(1 row)

select ('aa=>b, c=>d , b=>16'::hstore->'gg') is null;
 ?column? 
----------
 t
(1 row)

select ('aa=>NULL, c=>d , b=>16'::hstore->'aa') is null;
 ?column? 
----------
 t
(1 row)

select ('aa=>"NULL", c=>d , b=>16'::hstore->'aa') is null;
 ?column? 
----------
 f
(1 row)

-- -> array operator
select 'aa=>"NULL", c=>d , b=>16'::hstore -> ARRAY['aa','c'];
  ?column?  
------------
 {"NULL",d}
(1 row)

select 'aa=>"NULL", c=>d , b=>16'::hstore -> ARRAY['c','aa'];
  ?column?  
------------
 {d,"NULL"}
(1 row)

select 'aa=>NULL, c=>d , b=>16'::hstore -> ARRAY['aa','c',null];
  ?column?  
---------------
 {NULL,d,NULL}
(1 row)

select 'aa=>1, c=>3, b=>2, d=>4'::hstore -> ARRAY[['b','d'],['aa','c']];
  ?column?  
---------------
 {{2,4},{1,3}}
(1 row)

-- exists/defined
select exist('a=>NULL, b=>qq', 'a');
 exist 
-------
 t
(1 row)

select exist('a=>NULL, b=>qq', 'b');
 exist 
-------
 t
(1 row)

select exist('a=>NULL, b=>qq', 'c');
 exist 
-------
 f
(1 row)

select exist('a=>"NULL", b=>qq', 'a');
 exist 
-------
 t
(1 row)

select defined('a=>NULL, b=>qq', 'a');
 defined 
---------
 f
(1 row)

select defined('a=>NULL, b=>qq', 'b');
 defined 
---------
 t
(1 row)

select defined('a=>NULL, b=>qq', 'c');
 defined 
---------
 f
(1 row)

select defined('a=>"NULL", b=>qq', 'a');
 defined 
---------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ? 'a';
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ? 'b';
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ? 'c';
 ?column? 
----------
 f
(1 row)

select hstore 'a=>"NULL", b=>qq' ? 'a';
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ?| ARRAY['a','b'];
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ?| ARRAY['b','a'];
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ?| ARRAY['c','a'];
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ?| ARRAY['c','d'];
 ?column? 
----------
 f
(1 row)

select hstore 'a=>NULL, b=>qq' ?| '{}'::text[];
 ?column? 
----------
 f
(1 row)

select hstore 'a=>NULL, b=>qq' ?& ARRAY['a','b'];
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ?& ARRAY['b','a'];
 ?column? 
----------
 t
(1 row)

select hstore 'a=>NULL, b=>qq' ?& ARRAY['c','a'];
 ?column? 
----------
 f
(1 row)

select hstore 'a=>NULL, b=>qq' ?& ARRAY['c','d'];
 ?column? 
----------
 f
(1 row)

select hstore 'a=>NULL, b=>qq' ?& '{}'::text[];
 ?column? 
----------
 f
(1 row)

-- delete 
select delete('a=>1 , b=>2, c=>3'::hstore, 'a');
       delete       
--------------------
 "b"=>"2", "c"=>"3"
(1 row)

select delete('a=>null , b=>2, c=>3'::hstore, 'a');
       delete       
--------------------
 "b"=>"2", "c"=>"3"
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, 'b');
       delete       
--------------------
 "a"=>"1", "c"=>"3"
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, 'c');
       delete       
--------------------
 "a"=>"1", "b"=>"2"
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, 'd');
            delete            
------------------------------
 "a"=>"1", "b"=>"2", "c"=>"3"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - 'a'::text;
      ?column?      
--------------------
 "b"=>"2", "c"=>"3"
(1 row)

select 'a=>null , b=>2, c=>3'::hstore - 'a'::text;
      ?column?      
--------------------
 "b"=>"2", "c"=>"3"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - 'b'::text;
      ?column?      
--------------------
 "a"=>"1", "c"=>"3"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - 'c'::text;
      ?column?      
--------------------
 "a"=>"1", "b"=>"2"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - 'd'::text;
           ?column?           
------------------------------
 "a"=>"1", "b"=>"2", "c"=>"3"
(1 row)

-- delete (array)
select delete('a=>1 , b=>2, c=>3'::hstore, ARRAY['d','e']);
            delete            
------------------------------
 "a"=>"1", "b"=>"2", "c"=>"3"
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, ARRAY['d','b']);
       delete       
--------------------
 "a"=>"1", "c"=>"3"
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, ARRAY['a','c']);
  delete  
----------
 "b"=>"2"
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, ARRAY[['b'],['c'],['a']]);
 delete 
--------
 
(1 row)

select delete('a=>1 , b=>2, c=>3'::hstore, '{}'::text[]);
            delete            
------------------------------
 "a"=>"1", "b"=>"2", "c"=>"3"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - ARRAY['d','e'];
           ?column?           
------------------------------
 "a"=>"1", "b"=>"2", "c"=>"3"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - ARRAY['d','b'];
      ?column?      
--------------------
 "a"=>"1", "c"=>"3"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - ARRAY['a','c'];
 ?column? 
----------
 "b"=>"2"
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - ARRAY[['b'],['c'],['a']];
 ?column? 
----------
 
(1 row)

select 'a=>1 , b=>2, c=>3'::hstore - '{}'::text[];
           ?column?           
------------------------------
 "a"=>"1", "b"=>"2", "c"=>"3"
(1 row)

-- delete (hstore)
select delete('aa=>1 , b=>2, c=>3'::hstore, 'aa=>4, b=>2'::hstore);
       delete        
---------------------
 "c"=>"3", "aa"=>"1"
(1 row)

select delete('aa=>1 , b=>2, c=>3'::hstore, 'aa=>NULL, c=>3'::hstore);
       delete        
---------------------
 "b"=>"2", "aa"=>"1"
(1 row)

select delete('aa=>1 , b=>2, c=>3'::hstore, 'aa=>1, b=>2, c=>3'::hstore);
 delete 
--------
 
(1 row)

select delete('aa=>1 , b=>2, c=>3'::hstore, 'b=>2'::hstore);
       delete        
---------------------
 "c"=>"3", "aa"=>"1"
(1 row)

select delete('aa=>1 , b=>2, c=>3'::hstore, ''::hstore);
            delete             
-------------------------------
 "b"=>"2", "c"=>"3", "aa"=>"1"
(1 row)

select 'aa=>1 , b=>2, c=>3'::hstore - 'aa=>4, b=>2'::hstore;
      ?column?       
---------------------
 "c"=>"3", "aa"=>"1"
(1 row)

select 'aa=>1 , b=>2, c=>3'::hstore - 'aa=>NULL, c=>3'::hstore;
      ?column?       
---------------------
 "b"=>"2", "aa"=>"1"
(1 row)

select 'aa=>1 , b=>2, c=>3'::hstore - 'aa=>1, b=>2, c=>3'::hstore;
 ?column? 
----------
 
(1 row)

select 'aa=>1 , b=>2, c=>3'::hstore - 'b=>2'::hstore;
      ?column?       
---------------------
 "c"=>"3", "aa"=>"1"
(1 row)

select 'aa=>1 , b=>2, c=>3'::hstore - ''::hstore;
           ?column?            
-------------------------------
 "b"=>"2", "c"=>"3", "aa"=>"1"
(1 row)

-- ||
select 'aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>f';
                 ?column?                  
-------------------------------------------
 "b"=>"g", "aa"=>"1", "cq"=>"l", "fg"=>"f"
(1 row)

select 'aa=>1 , b=>2, cq=>3'::hstore || 'aq=>l';
                 ?column?                  
-------------------------------------------
 "b"=>"2", "aa"=>"1", "aq"=>"l", "cq"=>"3"
(1 row)

select 'aa=>1 , b=>2, cq=>3'::hstore || 'aa=>l';
            ?column?            
--------------------------------
 "b"=>"2", "aa"=>"l", "cq"=>"3"
(1 row)

select 'aa=>1 , b=>2, cq=>3'::hstore || '';
            ?column?            
--------------------------------
 "b"=>"2", "aa"=>"1", "cq"=>"3"
(1 row)

select ''::hstore || 'cq=>l, b=>g, fg=>f';
            ?column?            
--------------------------------
 "b"=>"g", "cq"=>"l", "fg"=>"f"
(1 row)

-- =>
select 'a=>g, b=>c'::hstore || ( 'asd'=>'gf' );
            ?column?             
---------------------------------
 "a"=>"g", "b"=>"c", "asd"=>"gf"
(1 row)

select 'a=>g, b=>c'::hstore || ( 'b'=>'gf' );
      ?column?       
---------------------
 "a"=>"g", "b"=>"gf"
(1 row)

select 'a=>g, b=>c'::hstore || ( 'b'=>'NULL' );
       ?column?        
-----------------------
 "a"=>"g", "b"=>"NULL"
(1 row)

select 'a=>g, b=>c'::hstore || ( 'b'=>NULL );
      ?column?       
---------------------
 "a"=>"g", "b"=>NULL
(1 row)

-- => arrays
select ARRAY['a','b','asd'] => ARRAY['g','h','i'];
            ?column?            
--------------------------------
 "a"=>"g", "b"=>"h", "asd"=>"i"
(1 row)

select ARRAY['a','b','asd'] => ARRAY['g','h',NULL];
            ?column?             
---------------------------------
 "a"=>"g", "b"=>"h", "asd"=>NULL
(1 row)

select ARRAY['z','y','x'] => ARRAY['1','2','3'];
           ?column?           
------------------------------
 "x"=>"3", "y"=>"2", "z"=>"1"
(1 row)

select ARRAY['aaa','bb','c','d'] => ARRAY[null::text,null,null,null];
                   ?column?                    
-----------------------------------------------
 "c"=>NULL, "d"=>NULL, "bb"=>NULL, "aaa"=>NULL
(1 row)

select hstore 'aa=>1, b=>2, c=>3' => ARRAY['g','h','i'];
 ?column? 
----------
 
(1 row)

select hstore 'aa=>1, b=>2, c=>3' => ARRAY['c','b'];
      ?column?      
--------------------
 "b"=>"2", "c"=>"3"
(1 row)

select hstore 'aa=>1, b=>2, c=>3' => ARRAY['aa','b'];
      ?column?       
---------------------
 "b"=>"2", "aa"=>"1"
(1 row)

select hstore 'aa=>1, b=>2, c=>3' => ARRAY['c','b','aa'];
           ?column?            
-------------------------------
 "b"=>"2", "c"=>"3", "aa"=>"1"
(1 row)

-- records
select hstore(v) from (values (1, 'foo', 1.2, 3::float8)) v(a,b,c,d);
                     hstore                     
------------------------------------------------
 "f1"=>"1", "f2"=>"foo", "f3"=>"1.2", "f4"=>"3"
(1 row)

create table testhstore1 (a integer, b text, c numeric, d float8);
insert into testhstore1 values (1, 'foo', 1.2, 3::float8);
select hstore(v) from testhstore1 v;
                   hstore                   
--------------------------------------------
 "a"=>"1", "b"=>"foo", "c"=>"1.2", "d"=>"3"
(1 row)

select populate_record(v, ('c' => '3.45')) from testhstore1 v;
 populate_record 
-----------------
 (1,foo,3.45,3)
(1 row)

select populate_record(v, ('d' => '3.45')) from testhstore1 v;
 populate_record  
------------------
 (1,foo,1.2,3.45)
(1 row)

select populate_record(v, ('c' => null)) from testhstore1 v;
 populate_record 
-----------------
 (1,foo,,3)
(1 row)

select populate_record(v, ('b' => 'foo') || ('a' => '123')) from testhstore1 v;
 populate_record 
-----------------
 (123,foo,1.2,3)
(1 row)

select populate_record(null::testhstore1, ('c' => '3.45') || ('a' => '123')) from testhstore1 v;
 populate_record 
-----------------
 (123,,3.45,)
(1 row)

select v #= ('c' => '3.45') from testhstore1 v;
    ?column?    
----------------
 (1,foo,3.45,3)
(1 row)

select v #= ('d' => '3.45') from testhstore1 v;
     ?column?     
------------------
 (1,foo,1.2,3.45)
(1 row)

select v #= ('c' => null) from testhstore1 v;
  ?column?  
------------
 (1,foo,,3)
(1 row)

select v #= (('b' => 'foo') || ('a' => '123')) from testhstore1 v;
    ?column?     
-----------------
 (123,foo,1.2,3)
(1 row)

select null::testhstore1 #= (('c' => '3.45') || ('a' => '123')) from testhstore1 v;
   ?column?   
--------------
 (123,,3.45,)
(1 row)

select v #= h from testhstore1 v, (values (hstore 'a=>123',1),('b=>foo,c=>3.21',2),('a=>null',3),('e=>blah',4)) x(h,i) order by i;
    ?column?     
-----------------
 (123,foo,1.2,3)
 (1,foo,3.21,3)
 (,foo,1.2,3)
 (1,foo,1.2,3)
(4 rows)

-- keys/values
select akeys('aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>f');
    akeys     
--------------
 {b,aa,cq,fg}
(1 row)

select akeys('""=>1');
 akeys 
-------
 {""}
(1 row)

select akeys('');
 akeys 
-------
 {}
(1 row)

select avals('aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>f');
   avals   
-----------
 {g,1,l,f}
(1 row)

select avals('aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>NULL');
   avals    
--------------
 {g,1,l,NULL}
(1 row)

select avals('""=>1');
 avals 
-------
 {1}
(1 row)

select avals('');
 avals 
-------
 {}
(1 row)

select * from skeys('aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>f');
 skeys 
-------
 b
 aa
 cq
 fg
(4 rows)

select * from skeys('""=>1');
 skeys 
-------
 
(1 row)

select * from skeys('');
 skeys 
-------
(0 rows)

select * from svals('aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>f');
 svals 
-------
 g
 1
 l
 f
(4 rows)

select *, svals is null from svals('aa=>1 , b=>2, cq=>3'::hstore || 'cq=>l, b=>g, fg=>NULL');
 svals | ?column? 
-------+----------
 g     | f
 1     | f
 l     | f
       | t
(4 rows)

select * from svals('""=>1');
 svals 
-------
 1
(1 row)

select * from svals('');
 svals 
-------
(0 rows)

select * from each('aaa=>bq, b=>NULL, ""=>1 ');
 key | value 
-----+-------
     | 1
 b   | 
 aaa | bq
(3 rows)

-- @>
select 'a=>b, b=>1, c=>NULL'::hstore @> 'a=>b';
 ?column? 
----------
 t
(1 row)

select 'a=>b, b=>1, c=>NULL'::hstore @> 'a=>b, c=>NULL';
 ?column? 
----------
 t
(1 row)

select 'a=>b, b=>1, c=>NULL'::hstore @> 'a=>b, g=>NULL';
 ?column? 
----------
 f
(1 row)

select 'a=>b, b=>1, c=>NULL'::hstore @> 'g=>NULL';
 ?column? 
----------
 f
(1 row)

select 'a=>b, b=>1, c=>NULL'::hstore @> 'a=>c';
 ?column? 
----------
 f
(1 row)

select 'a=>b, b=>1, c=>NULL'::hstore @> 'a=>b';
 ?column? 
----------
 t
(1 row)

select 'a=>b, b=>1, c=>NULL'::hstore @> 'a=>b, c=>q';
 ?column? 
----------
 f
(1 row)

CREATE TABLE testhstore (h hstore);
\copy testhstore from 'data/hstore.data'
select count(*) from testhstore where h @> 'wait=>NULL';
 count 
-------
     1
(1 row)

select count(*) from testhstore where h @> 'wait=>CC';
 count 
-------
    15
(1 row)

select count(*) from testhstore where h @> 'wait=>CC, public=>t';
 count 
-------
     2
(1 row)

select count(*) from testhstore where h ? 'public';
 count 
-------
   194
(1 row)

select count(*) from testhstore where h ?| ARRAY['public','disabled'];
 count 
-------
   337
(1 row)

select count(*) from testhstore where h ?& ARRAY['public','disabled'];
 count 
-------
    42
(1 row)

create index hidx on testhstore using gist(h);
set enable_seqscan=off;
select count(*) from testhstore where h @> 'wait=>NULL';
 count 
-------
     1
(1 row)

select count(*) from testhstore where h @> 'wait=>CC';
 count 
-------
    15
(1 row)

select count(*) from testhstore where h @> 'wait=>CC, public=>t';
 count 
-------
     2
(1 row)

select count(*) from testhstore where h ? 'public';
 count 
-------
   194
(1 row)

select count(*) from testhstore where h ?| ARRAY['public','disabled'];
 count 
-------
   337
(1 row)

select count(*) from testhstore where h ?& ARRAY['public','disabled'];
 count 
-------
    42
(1 row)

drop index hidx;
create index hidx on testhstore using gin (h);
set enable_seqscan=off;
select count(*) from testhstore where h @> 'wait=>NULL';
 count 
-------
     1
(1 row)

select count(*) from testhstore where h @> 'wait=>CC';
 count 
-------
    15
(1 row)

select count(*) from testhstore where h @> 'wait=>CC, public=>t';
 count 
-------
     2
(1 row)

select count(*) from testhstore where h ? 'public';
 count 
-------
   194
(1 row)

select count(*) from testhstore where h ?| ARRAY['public','disabled'];
 count 
-------
   337
(1 row)

select count(*) from testhstore where h ?& ARRAY['public','disabled'];
 count 
-------
    42
(1 row)

select count(*) from (select (each(h)).key from testhstore) as wow ;
 count 
-------
  4781
(1 row)

select key, count(*) from (select (each(h)).key from testhstore) as wow group by key order by count desc, key;
    key    | count 
-----------+-------
 line      |   884
 query     |   207
 pos       |   203
 node      |   202
 space     |   197
 status    |   195
 public    |   194
 title     |   190
 wait      |   190
 org       |   189
 user      |   189
 coauthors |   188
 disabled  |   185
 indexed   |   184
 cleaned   |   180
 bad       |   179
 date      |   179
 world     |   176
 state     |   172
 subtitle  |   169
 auth      |   168
 abstract  |   161
(22 rows)

-- btree opclass transitivity
set enable_seqscan=on;
create table testhstore2 as select * from testhstore limit 50;
create function bool_imp(boolean,boolean) returns boolean language sql as $f$ select (not $1) or $2; $f$;
create operator => (leftarg = boolean, rightarg = boolean, procedure = bool_imp);
select every( ((h1.h #># h2.h AND h2.h #># h3.h) => (h1.h #># h3.h)) AND ((h1.h #>=# h2.h AND h2.h #>=# h3.h) => (h1.h #>=# h3.h)) AND ((h1.h #<# h2.h AND h2.h #<# h3.h) => (h1.h #<# h3.h)) AND ((h1.h #<=# h2.h AND h2.h #<=# h3.h) => (h1.h #<=# h3.h)) AND ((h1.h = h2.h AND h2.h = h3.h) => (h1.h = h3.h)) ) from testhstore2 h1, testhstore2 h2, testhstore2 h3;
 every 
-------
 t
(1 row)

-- sort/hash
select count(distinct h) from testhstore;
 count 
-------
   885
(1 row)

set enable_hashagg = false;
select count(*) from (select h from (select * from testhstore union all select * from testhstore) hs group by h) hs2;
 count 
-------
   885
(1 row)

set enable_hashagg = true;
set enable_sort = false;
select count(*) from (select h from (select * from testhstore union all select * from testhstore) hs group by h) hs2;
 count 
-------
   885
(1 row)

set enable_sort = true;
-- btree
drop index hidx;
create index hidx on testhstore using btree (h);
set enable_seqscan=off;
select count(*) from testhstore where h #># 'p=>1';
 count 
-------
   125
(1 row)

select count(*) from testhstore where h = 'pos=>98, line=>371, node=>CBA, indexed=>t';
 count 
-------
     1
(1 row)

